### PR TITLE
New content analyzis

### PR DESCRIPTION
**Pull request by JalilArfaoui**
TLDR; Code is not final, this is more a proof of concept that we can find the currently displayed amazon product title.

Objective was to use accessibility service to extract current DOM from Chrome, to be able to apply XPath rule on live content.

After a lot of trials and reading, I failed to meet this main goal.

What I managed to do is to deep analyze the DOM, but exposed as a tree of `AccessibilityNodeInfo`. Tag names and classes are not exposed (every tag is exposed as an Android component, `View` and so on …), but we seem to have the full hierarchy with ids and text contents.

So, I manage to build a poc/feature that, whenever a product page is loaded in amazon (or when app becomes visible), irrespective of the URL, we can find the product title :

```kotlin
 val webview = findWebview(root)
    if (webview != null) {
      val titleExpanderContent = findById(webview, "titleExpanderContent")
      if (titleExpanderContent != null) {
        val titleView = findHeading(titleExpanderContent)
        val title = titleView?.text ?: titleView?.contentDescription
        Log.d(TAG, "Found Amazon page title : $title")
      }
    }
```

gives

```bash
D/Accessibility: Event : TYPE_WINDOW_CONTENT_CHANGED, Package: com.google.android.apps.nexuslauncher, Source: null
D/Accessibility: Active window packageName : com.android.chrome, className: android.widget.FrameLayout
D/Accessibility: Found Amazon page title : 6S Casque Bluetooth sans Fil, écouteurs stéréo sans Fil stéréo Pliables Hi-FI Écouteurs avec Microphone intégré, Micro SD/TF, FM pour iPhone/Samsung/iPad/PC (Or Noir)
```

Now, we need to discuss what’s the best strategy to replace or full XPath strategy :
 - Keep current strategy (fetch current URL document in background and apply XPath on it)
 - Replace with a mechanism to find more specific informations on more specific websites
 - Both
 -  …
 
 Used documentation and references :
 - https://stuff.mit.edu/afs/sipb/project/android/docs/guide/topics/ui/accessibility/services.html
 - https://medium.com/nerd-for-tech/track-web-browser-usage-in-android-using-accessibility-service-800bfa2745d2
 - https://stackoverflow.com/questions/33318083/how-to-get-webview-from-accessibilitynodeinfo
 - https://groups.google.com/a/chromium.org/g/chromium-dev/c/2VC16XswAaI
 - https://stackoverflow.com/questions/7282789/is-there-any-way-to-get-access-to-dom-structure-in-androids-webview
 - https://stackoverflow.com/questions/40522043/how-to-access-html-content-of-accessibilitynodeinfo-of-a-webview-element-using-a
 - https://stackoverflow.com/questions/65326148/why-accessibilityservice-failed-to-retrieve-content-of-a-webview-but-works-prop
 - https://github.com/google/talkback
 - https://github.com/chromium/chromium/blob/master/content/public/android/java/src/org/chromium/content/browser/accessibility/WebContentsAccessibilityImpl.java
 - https://www.py4u.net/discuss/630533
 - https://stackoverflow.com/questions/10634908/accessibility-and-android-webview
 - https://stackoverflow.com/questions/36793154/accessibilityservice-not-returning-view-ids

### CODE REVIEW

**Comment by zhinu** [comment]
Rules (of new system):
Hierarchy is understood
node ids can be found
node names cannot be found (except for h1 h2 ...)

**Comment by JalilArfaoui** [comment]
I’ve just pushed a big refacto in order to take the direction decided in https://github.com/dis-moi/mobile/issues/234 and https://github.com/dis-moi/backend/issues/448

## Technical context :

###  In current master version, simplified

- We use accessibility services, Kotlin side, to detect URL, and we send all URLs to React Native HeadlessTask
- In RN HeadlessTask we first fetch all matching contexts, and if current url matches and we have an xpath condition, we send it back to a Kotlin XPath module
- In the Kotlin XPath module, we fetch current url and try on apply xpath on fetched html, and then we send the response back to RN HeadlessTask **<- Here we don’t have access to accessiblity informations anymore**
- In RN HeadlessTask, we have some last checks (notice hasn’t been deleted, …) and then we call a Kotlin UI service that manages all the floating UI, to display found notices 

### Version proposed here 

 The main work I did here (in https://github.com/dis-moi/mobile/pull/233/commits/8ac371692daea97e8ab230cbe9330a26eee61502 in particular) was to be able to analyze page content given a list of matching contexts to check. 
 
 The thing is, because current content analysis (XPath) is done later and only given an URL, I could not change the strategy in a drop-in manner … 
 
 I had to bring back most of the context matching process together, in the accessibility service, so that we can push screen content analysis further when needed 

So, now, the `BackgroundService` (that deserves a better name) does the following :  

- Fetch all matching contexts (only once, will be happen regularly)
- Every time screen change (well, with a 500 ms debounce), detect application, and if supported, current URL
- Find all corresponding matching contexts
- **<-- If any has content condition, we’re ready to analyze here**
- Then we send all Ids to the Floating UI service I adapted for this 

I do not use RN Headless task anymore, but it’s still there because I did not port all the functionality yet. Same for some other files, as FloatingModule, FloatingCoordinator, XPathModule …

## Functional

### What this version (also) brings

- Firefox support
- Background matching context fetch
- Clean bubble show/hide/update on home / app switch / tab switch / keyboard type

### What this version misses 

A lot of things need to be «plugged back» :

- cannot dismiss bubble notification 
- cannot delete message
- cannot close message window
- do not take subscribed contributors into account
- links in messages not working
- message window does not hide on screen change

## Technical debt

I’ve let a lot of `TODO` comments … 

- Matching contexts and notices fetch should be moved to a model/repository
- Avoid `observeForever`
- Use proper data types
- Fix typefaces
- Sentry could be addde
- …

File: android/app/src/main/java/com/dismoi/scout/floating/FloatingService.kt [hunk 1]
```diff
@@ -4,139 +4,252 @@ import android.app.Service
...
+
+  override fun onBind(intent: Intent?): IBinder {
     return binder
   }
 
-  override fun onUnbind(intent: Intent): Boolean {
-    return super.onUnbind(intent)
+  // TODO Proper typed an mapped object !
+  @RequiresApi(Build.VERSION_CODES.N)
+  fun showNotices(noticesIds: List<Int>) {
```
Comment (lutangar) [inline android/app/src/main/java/com/dismoi/scout/floating/FloatingService.kt:69]: Maybe some part of this function's logic could be handled in the `ContextMatcherService` to reduce the responsabilty of this service to *presentation* only.

File: android/app/src/main/java/com/dismoi/scout/floating/FloatingService.kt [hunk 2]
```diff
@@ -4,139 +4,252 @@ import android.app.Service
...
+    showView(trashView, trashLayoutParams)
+  }
 
+  private fun hideTrash() {
+    removeView(trashView)
   }
 
-  fun removeMessage(message: Message?) {
-    getWindowManager()!!.removeView(message)
+  // TODO To be moved to a repository
```
Comment (lutangar) [inline android/app/src/main/java/com/dismoi/scout/floating/FloatingService.kt:228]: Definitely!

File: android/app/src/main/java/com/dismoi/scout/floating/Manager.kt [hunk 3]
```diff
@@ -6,21 +6,18 @@ import android.content.Intent
 import android.content.ServiceConnection
 import android.os.IBinder
 import android.util.Log
-import com.dismoi.scout.floating.layout.Bubble
-import com.dismoi.scout.floating.layout.Layout
-import com.dismoi.scout.floating.layout.Message
 import com.dismoi.scout.floating.FloatingService.FloatingServiceBinder
 
 class Manager private constructor(private val context: Context) {
```
Comment (lutangar) [inline android/app/src/main/java/com/dismoi/scout/floating/Manager.kt:11]: A class called a `Manager` don't convey much meaning, this is sooo confusing. But I know you know ;)

File: android/app/src/main/java/com/dismoi/scout/floating/layout/Message.kt [hunk 4]
```diff
@@ -1,20 +1,157 @@
...
+        val avatarView = noticeView.findViewById<ImageView>(R.id.contributorProfile)
+        val avatarBitmap =
+          BitmapFactory.decodeStream(contributorAvatarUrl.openConnection().getInputStream())
+        avatarView.setImageBitmap(avatarBitmap)
+      }
+    }
+
+    liveNotice.observeForever(noticeObserver) // TODO we should link to a LifecycleOwner
+
+    noticeView
```
Comment (lutangar) [inline android/app/src/main/java/com/dismoi/scout/floating/layout/Message.kt:104]: Looks like a wip, but might be a language feature I'm not aware of.

File: android/app/src/main/java/com/dismoi/scout/floating/layout/Message.kt [hunk 5]
```diff
@@ -1,20 +1,157 @@
...
+      field = notices
+    }
+
+  private var inflater: LayoutInflater = context.getSystemService(Service.LAYOUT_INFLATER_SERVICE) as LayoutInflater
+
+  // TODO Fix fonts usage
+  //    val type = Typeface.createFromAsset(context.assets, "fonts/Helvetica.ttf")
+  //    val typeBold = Typeface.createFromAsset(context.assets, "fonts/Helvetica-Bold.ttf")
+
+  private var carouselListener = ViewListener { position ->
```
Comment (lutangar) [inline android/app/src/main/java/com/dismoi/scout/floating/layout/Message.kt:46]: This method is soooo long, this would deserve to be break down in multiple function to be better understood.

**Review by lutangar — CHANGES_REQUESTED** [review]

**Comment by lutangar** [comment]
@JalilArfaoui You're describing interesting domain logic and mentionning resources in the PR description which would both  deserve to appears somewhere in a markdown!

File: android/app/src/main/java/com/dismoi/scout/accessibility/BackgroundService.kt [hunk 6]
```diff
@@ -1,193 +1,229 @@
...
-        _packageName = packageName
-        handler.post(runnableCode)
-        return
+        show(noticesId)
+      } else {
+        hide()
       }
+    } catch (e: NoUrlInBrowserException)
+    {
+      // When page is scrolled, the URL is hidden, but we may still be on the same page …
```
[the remark this answers is not in the document]
Comment (JalilArfaoui) [reply android/app/src/main/java/com/dismoi/scout/accessibility/BackgroundService.kt:149]: It **will** happen very frequently, and it’s normal. We just need to not act on it.

[the remark this answers is not in the document]
Comment (JalilArfaoui) [reply android/app/src/main/java/com/dismoi/scout/accessibility/BackgroundService.kt:149]: And events are already debounced … I just don’t think a log adds value …

File: android/app/src/main/java/com/dismoi/scout/accessibility/BackgroundService.kt [hunk 7]
```diff
@@ -1,193 +1,229 @@
...
+
+      Log.d(TAG, "current URL is $currentUrl")
+
+      val matchingContexts = getContextsMatchingUrl(currentUrl)
+
+      Log.d(TAG, "found ${matchingContexts.size} contexts matching current url")
+
+      if (matchingContexts.size > 0) {
+
+        val noticesId =
```
[the remark this answers is not in the document]
Comment (JalilArfaoui) [reply android/app/src/main/java/com/dismoi/scout/accessibility/BackgroundService.kt:133]: Done …

File: android/app/src/main/java/com/dismoi/scout/accessibility/BackgroundService.kt [hunk 8]
```diff
@@ -1,193 +1,229 @@
...
     }
   }
 
-  override fun onInterrupt() {
-    sendEventFromAccessibilityServicePermission("false")
+  // TODO To be moved to repository
+  private fun fetchMatchingContexts() {
+    val matchingContextsEndpoint = "https://notices.bulles.fr/api/v3/matching-contexts" // TODO to be moved to env configuration
+    val request = Request.Builder().url(matchingContextsEndpoint).build()
+    val client = OkHttpClient()
```
[the remark this answers is not in the document]
Comment (JalilArfaoui) [reply android/app/src/main/java/com/dismoi/scout/accessibility/BackgroundService.kt:179]: Yes, I wanted to not put everything in this single PR … This will totally be refactored as part of https://github.com/dis-moi/mobile/issues/173#issuecomment-886119063

File: android/app/src/main/java/com/dismoi/scout/accessibility/BackgroundService.kt [hunk 9]
```diff
@@ -1,193 +1,229 @@
...
+    runBlocking { launch {
+      eventsChannel.send(event)
+    } }
   }
 
-  private fun isWindowChangeEvent(event: AccessibilityEvent): Boolean {
-    return AccessibilityEvent.eventTypeToString(event.eventType).contains("WINDOW")
-  }
+  @RequiresApi(Build.VERSION_CODES.N)
+  private fun matchContext(event: AccessibilityEvent) {
```
[the remark this answers is not in the document]
Comment (JalilArfaoui) [reply android/app/src/main/java/com/dismoi/scout/accessibility/BackgroundService.kt:89]: I tried to differentiate :
- `MatchingContext` : The description of a context that should match
- `matchContext` : The action of checking wether a MatchingContext matches a specific context … Would you prefer `checkContext` ? `checkIfContextIsMatching` ?

[the remark this answers is not in the document]
Comment (JalilArfaoui) [reply android/app/src/main/java/com/dismoi/scout/accessibility/BackgroundService.kt:89]: And, yes, `BackgroundService` should totally be renamed … Just tried not to do everything here …

File: android/app/src/main/java/com/dismoi/scout/browser/Helpers.kt [hunk 10]
```diff
@@ -0,0 +1,142 @@
+package com.dismoi.scout.browser
+
+import android.os.Build
+import android.util.Log
+import android.view.accessibility.AccessibilityNodeInfo
+import androidx.annotation.RequiresApi
+
+object Helpers {
```
[the remark this answers is not in the document]
Comment (JalilArfaoui) [reply android/app/src/main/java/com/dismoi/scout/browser/Helpers.kt:8]: OOP vs functional style :-)

About the name, I figured the package being `browser.Helpers` it was sufficient … but other folks seem to agree with you on this matter : https://stackoverflow.com/questions/13377766/should-java-class-name-duplicate-package-name

[the remark this answers is not in the document]
Comment (JalilArfaoui) [reply android/app/src/main/java/com/dismoi/scout/browser/Helpers.kt:8]: Created #244 to follow-up and unblock

File: android/app/src/main/java/com/dismoi/scout/floating/FloatingService.kt (hunk 2 above)
Comment (JalilArfaoui) [reply android/app/src/main/java/com/dismoi/scout/floating/FloatingService.kt:228]: #173

File: android/app/src/main/java/com/dismoi/scout/floating/Manager.kt (hunk 3 above)
Comment (JalilArfaoui) [reply android/app/src/main/java/com/dismoi/scout/floating/Manager.kt:11]: And I know you know I know !

Comment (JalilArfaoui) [reply android/app/src/main/java/com/dismoi/scout/floating/Manager.kt:11]: (And actually, you only see deletion here because I’m progressively getting rid of it …)

File: android/app/src/main/java/com/dismoi/scout/floating/layout/Message.kt (hunk 4 above)
Comment (JalilArfaoui) [reply android/app/src/main/java/com/dismoi/scout/floating/layout/Message.kt:104]: Yep, last line is the return value !

File: android/app/src/main/java/com/dismoi/scout/floating/layout/Message.kt (hunk 5 above)
Comment (JalilArfaoui) [reply android/app/src/main/java/com/dismoi/scout/floating/layout/Message.kt:46]: Totally !

Comment (JalilArfaoui) [reply android/app/src/main/java/com/dismoi/scout/floating/layout/Message.kt:46]: #245

File: android/app/src/main/java/com/dismoi/scout/floating/FloatingService.kt (hunk 1 above)
Comment (JalilArfaoui) [reply android/app/src/main/java/com/dismoi/scout/floating/FloatingService.kt:69]: We do not match here, but we do still have fetching that should be moved to repository … Again, will be done after #173

**Comment by JalilArfaoui** [comment]
> @JalilArfaoui You're describing interesting domain logic and mentionning resources in the PR description which would both deserve to appears somewhere in a markdown!

You’re right @lutangar … I’ve added docs/Architecture.md that tries to sum up
